homeassistant/components/mobile_app/__init__.py

```diff
@@ -0,0 +1,326 @@
+        return json_response({"message": "Event {} fired.".format(event_type)})
+    elif webhook_type == WEBHOOK_TYPE_RENDER_TEMPLATE:
+        try:
+            tpl = template.Template(data[ATTR_TEMPLATE], hass)
+            rendered = tpl.async_render(data.get(ATTR_TEMPLATE_VARIABLES))
+            return json_response({"rendered": rendered})
+        except (ValueError, TemplateError) as ex:
+            return json_response("Error rendering template: {}".format(ex),
+                                 status=HTTP_BAD_REQUEST)
+
+    elif webhook_type == WEBHOOK_TYPE_UPDATE_LOCATION:
+        await hass.services.async_call(DEVICE_TRACKER_DOMAIN,
+                                       DEVICE_TRACKER_SEE, data)
+        return json_response([])
+
+
+def supports_encryption():
+    """Test if we support encryption."""
+    try:
+        import nacl   # noqa pylint: disable=unused-import
+        return True
+    except OSError:
+        return False
+
+
+def device_for_webhook_id(hass, webhook_id):
+    """Return the device name for the webhook ID."""
+    for device_name, device in hass.data[DOMAIN].items():
+        if device.get(CONF_WEBHOOK_ID) == webhook_id:
+            return device
+    return None
+
+
+async def async_setup(hass, config):
+    """Set up the mobile app component."""
+    conf = config.get(DOMAIN)
+
+    store = hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
+    app_config = await store.async_load()
+    if app_config is None:
+        app_config = {}
+
+    hass.data[DOMAIN] = app_config
+
+    for name, device in app_config.items():
+        if CONF_WEBHOOK_ID in device:
+            webhook.async_register(hass, DOMAIN, 'Mobile app',
+                                   device[CONF_WEBHOOK_ID], handle_webhook)
+
+    if conf is not None:
+        hass.async_create_task(hass.config_entries.flow.async_init(
+            DOMAIN, context={'source': config_entries.SOURCE_IMPORT}))
+
+    return True
+
+
+async def async_setup_entry(hass, entry):
+    """Set up an mobile app entry."""
+    hass.http.register_view(RegisterDeviceView())
+
+    return True
+
+
+class RegisterDeviceView(HomeAssistantView):
+    """A view that accepts device registration requests."""
+
+    url = '/api/mobile_app/register'
+    name = 'api:mobile_app:register'
+
+    def __init__(self):
+        """Initialize the view."""
+
+    async def post(self, request):
+        """Handle the POST request for device registration."""
+        try:
+            req_data = await request.json()
+        except ValueError:
+            return self.json_message("Invalid JSON", HTTP_BAD_REQUEST)
+
+        hass = request.app['hass']
+
+        try:
+            data = REGISTER_DEVICE_SCHEMA(req_data)
+        except vol.Invalid as ex:
+            return self.json_message(
+                vol.humanize.humanize_error(req_data, ex),
+                HTTP_BAD_REQUEST)
+
+        device_id = data[ATTR_DEVICE_ID]
+
+        if device_id in hass.data[DOMAIN]:
```

Comment:
This is weird, this makes the endpoint, accessed via POST, behave like a GET. If we want to support this, we should do this when handling the GET method. However, we should also verify that the user id matches.

Is there a use case for this behavior?

Reply:
Originally I had this returning a 409 but realized that we may want to allow apps to call this multiple times with the same payload in order to ensure they are still a valid entry in the system before making a webhook call.

Reply:
I don't like it, because we're now going to expose an encryption key to potentially a different app. If an app forgets, just let them re-register with a new ID.

Reply:
actually thinking here, should we even allow the app to pick a device ID  ? Why don't we generate one ? That way it's also never existing.

Reply:
The primary reason would be if you want to combine an app with existing `device_tracker` entry, e.g. iOS app + Unifi provided `device_tracker` updates

Reply:
We are going to deprecate device tracker merging in favor of merging in the person component.

Reply:
Okay. Do you have any specific suggestion/requirement on how to generate the device ID, or is `generate_secret` enough?

Reply:
@balloob Just realized another issue with not allowing the app to choose `device_id`: it could register devices all day with the same payload, unless we want to key off of `device_name` which I would **not** recommend, since the default name for an iOS device is `iPhone`, `iPad`, `iPod touch`, etc., so we are likely to hit conflicts.

Open to other suggestions.

Reply:
We need some kind of unique identifier. It can be left up to the apps to decide, but it needs to be there to ensure uniqueness. Will add a property named `fingerprint` to support this.